internal/server/v2/observation/calculation_util.go

```diff
@@ -183,29 +201,106 @@ func evalBinaryExpr(
 			}
 		}
 		if len(newOrderedFacets) > 0 {
-			if len(result.ByEntity) == 0 {
-				result.ByEntity = map[string]*pbv2.EntityObservation{}
+			result.ByEntity[entity] = &pbv2.EntityObservation{
+				OrderedFacets: newOrderedFacets,
 			}
+		}
+	}
+	return &intermediateResponse{variableObs: result}, nil
+}
+
+// Combine one VariableObservation with one constant using an operator token.
+func evalBinaryVariableConstantNodeExpr(
+	intermediate *pbv2.VariableObservation,
+	constant *float64,
+	iFirst bool, // Whether the intermediate response is the first response in the expression.
+	op token.Token,
+) (*intermediateResponse, error) {
+	result := &pbv2.VariableObservation{ByEntity: map[string]*pbv2.EntityObservation{}}
+	for entity, iEntityObs := range intermediate.ByEntity {
+		facets := iEntityObs.OrderedFacets
+		newOrderedFacets := []*pbv2.FacetObservation{}
+		for i := 0; i < len(facets); i++ {
+			newFacetId := facets[i].GetFacetId()
+			newPointStat := []*pb.PointStat{}
+			for _, obs := range facets[i].Observations {
+				iVal := obs.GetValue()
+				var val float64
+				var err error
+				if iFirst {
+					val, err = evalOp(iVal, *constant, op)
+				} else {
+					val, err = evalOp(*constant, iVal, op)
+				}
+				if err != nil {
+					return nil, err
+				}
+				newPointStat = append(newPointStat, &pb.PointStat{
+					Date:  obs.GetDate(),
+					Value: proto.Float64(val),
+				})
+			}
+			if len(newPointStat) > 0 {
+				newOrderedFacets = append(newOrderedFacets, &pbv2.FacetObservation{
+					FacetId:      newFacetId,
+					Observations: newPointStat,
+					EarliestDate: newPointStat[0].GetDate(),
+					LatestDate:   newPointStat[len(newPointStat)-1].GetDate(),
+					ObsCount:     int32(len(newPointStat)),
+				})
+			}
+		}
+		if len(newOrderedFacets) > 0 {
 			result.ByEntity[entity] = &pbv2.EntityObservation{
 				OrderedFacets: newOrderedFacets,
 			}
 		}
 	}
-	return result, nil
+	return &intermediateResponse{variableObs: result}, nil
+}
+
+// Combine two intermediateResponses using an operator token.
+func evalBinaryExpr(
+	x, y *intermediateResponse,
+	op token.Token,
+) (*intermediateResponse, error) {
+	if (x.variableObs != nil) && (y.variableObs != nil) {
```

Comment:
Does this assume that `x` and `y` will either include a `variableObs` or a `constantObs` but never both? And what happens if it does?

Reply:
They both shouldn't be set (since we only ever set one at a time), but we prefer `variableObs` over any `constantObs`. Added a comment clarifying this